include/msgpack/v1/adaptor/raw.hpp

```diff
@@ -34,7 +34,7 @@ struct raw_ref {
 
     bool operator== (const raw_ref& x) const
     {
-        return size == x.size && std::memcmp(ptr, x.ptr, size) == 0;
+        return size == x.size && (size == 0 || std::memcmp(ptr, x.ptr, size) == 0);
```

Comment:
Why size == 0 is needed?

In https://en.cppreference.com/w/cpp/string/byte/memcmp
> ​0​ if all count bytes of lhs and rhs are equal.



In https://en.cppreference.com/w/c/string/byte/memcmp
> Zero if lhs and rhs compare equal, or if count is zero.

I think that if size is 0 then return value is 0.
C notation is very straightforward. C++ notation is difficult to understand but I think that the meaning is the same.

Reply:
It is related topic that the `std::memcmp()` notation is follows:

https://en.cppreference.com/w/cpp/string/byte/memcpy
> If either dest or src is an invalid or null pointer, the behavior is undefined, even if count is zero.

Reply:
In https://en.cppreference.com/w/c/string/byte/memcmp

> The behavior is undefined if access occurs beyond the end of either object pointed to by lhs and rhs. **The behavior is undefined if either lhs or rhs is a null pointer.**

I actually found a configuration when this problem was caught with valgrind. To reproduce the problem, revert lines 37, 47 and 53 in `include/msgpack/v1/adaptor/raw.hpp` to their previous state (without checking size against 0) and run this (assuming current directory is repository root):

```
mkdir build-problem && cd build-problem
cmake -GNinja -DMSGPACK_BUILD_TESTS=ON -DMSGPACK_DEFAULT_API_VERSION=1 -DCMAKE_CXX_FLAGS="-g" ..
cmake --build . --target all
ctest -T memcheck
```

The output will contain:

```
      Start 23: raw
23/31 MemCheck #23: raw ..............................***Failed    1.58 sec

97% tests passed, 1 tests failed out of 31

Total Test time (real) = 177.14 sec

The following tests FAILED:
         23 - raw (Failed)
-- Processing memory checking output:
23/31 MemCheck: #23: raw ..............................   Defects: 1
MemCheck log files can be found here: (<#> corresponds to test number)
/home/kovdan01/Projects/msgpack-c/build-problem/Testing/Temporary/MemoryChecker.<#>.log
Memory checking results:
Uninitialized Memory Conditional - 1
        Errors while running CTest
Output from these tests are in: /home/kovdan01/Projects/msgpack-c/build-problem/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
```

MemoryChecker.23.log file contains the following:

```
==15705== Memcheck, a memory error detector
==15705== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==15705== Using Valgrind-3.17.0 and LibVEX; rerun with -h for copyright info
==15705== Command: /home/kovdan01/Projects/msgpack-c/build-problem/test/raw
==15705== Parent PID: 15704
==15705==
==15705== Conditional jump or move depends on uninitialised value(s)
==15705==    at 0x48479D0: bcmp (vg_replace_strmem.c:1129)
==15705==    by 0x12855A: msgpack::v1::type::raw_ref::operator==(msgpack::v1::type::raw_ref const&) const (raw.hpp:37)
==15705==    by 0x11AED7: pack_unpack_8_l::test_method() (raw.cpp:47)
==15705==    by 0x11A7B0: pack_unpack_8_l_invoker() (raw.cpp:33)
==15705==    by 0x12D2BB: boost::detail::function::void_function_invoker0<void (*)(), void>::invoke(boost::detail::function::function_buffer&) (function_template.hpp:117)
==15705==    by 0x48AA66D: ??? (in /usr/lib/libboost_unit_test_framework.so.1.76.0)
==15705==    by 0x48A8F5C: boost::execution_monitor::catch_signals(boost::function<int ()> const&) (in /usr/lib/libboost_unit_test_framework.so.1.76.0)
==15705==    by 0x48A8FE6: boost::execution_monitor::execute(boost::function<int ()> const&) (in /usr/lib/libboost_unit_test_framework.so.1.76.0)
==15705==    by 0x48A9098: boost::execution_monitor::vexecute(boost::function<void ()> const&) (in /usr/lib/libboost_unit_test_framework.so.1.76.0)
==15705==    by 0x48D78D6: boost::unit_test::unit_test_monitor_t::execute_and_translate(boost::function<void ()> const&, unsigned long) (in /usr/lib/libboost_unit_test_framework.so.1.76.0)
==15705==    by 0x48AF513: ??? (in /usr/lib/libboost_unit_test_framework.so.1.76.0)
==15705==    by 0x48AF7D7: ??? (in /usr/lib/libboost_unit_test_framework.so.1.76.0)
==15705==
==15705==
==15705== HEAP SUMMARY:
==15705==     in use at exit: 0 bytes in 0 blocks
==15705==   total heap usage: 3,372 allocs, 3,372 frees, 2,271,883 bytes allocated
==15705==
==15705== All heap blocks were freed -- no leaks are possible
==15705==
==15705== Use --track-origins=yes to see where uninitialised values come from
==15705== For lists of detected and suppressed errors, rerun with: -s
==15705== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

So I think that checking size against zero is required, otherwise it leads to UB when one of the pointers is nullptr.

Reply:
I think that it didn't happen before.
I wrote the following simple code.

```cpp
#include <cassert>
#include <cstring>

int main() {
    char* p1 = nullptr;
    char* p2 = nullptr;
    volatile std::size_t n = 0;
    assert(std::memcmp(p1, p2, n) == 0);
}
```

valgrind doesn't report error at least no option.

```
> clang++ -O3 memcmp.cpp && valgrind ./a.out

==135996== Memcheck, a memory error detector
==135996== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==135996== Using Valgrind-3.17.0 and LibVEX; rerun with -h for copyright info
==135996== Command: ./a.out
==135996==
==135996==
==135996== HEAP SUMMARY:
==135996==     in use at exit: 0 bytes in 0 blocks
==135996==   total heap usage: 1 allocs, 1 frees, 72,704 bytes allocated
==135996==
==135996== All heap blocks were freed -- no leaks are possible
==135996==
==135996== For lists of detected and suppressed errors, rerun with: -s
==135996== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

Could you tell me valgrined option and write a minimal code that reproduces the valgrined error?
I want to know why it happens.

Reply:
I removed size == 0 checking from raw.hpp and build on local environment and execute with valgrined. No errors are reported.
I use `valgrind-3.17.0` with no option.

Reply:
> Could you tell me valgrined option and write a minimal code that reproduces the valgrined error?
I want to know why it happens.

It seems that valgrind does not detect problems in your code sample, but the saninizer does:

```
> clang++ -O3 -fsanitize=address,undefined memcmp.cpp && ./a.out
memcmp.cpp:8:5: runtime error: null pointer passed as argument 1, which is declared to never be null
/usr/include/string.h:65:33: note: nonnull attribute specified here
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior memcmp.cpp:8:5 in 
memcmp.cpp:8:5: runtime error: null pointer passed as argument 2, which is declared to never be null
/usr/include/string.h:65:33: note: nonnull attribute specified here
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior memcmp.cpp:8:5 in 
```

BTW, my string.h header contains the following memcmp declaration, directly specifying that pointers must not be null:

```c++
/* Compare N bytes of S1 and S2.  */
extern int memcmp (const void *__s1, const void *__s2, size_t __n)
     __THROW __attribute_pure__ __nonnull ((1, 2));
```

> I use valgrind-3.17.0 with no option.

I have the same valgrind version. If you directly use valgrind (like `valgrind ctest`), please try to use `ctest -T memcheck` as mentioned above - these commands behave differently on my machine. Unfortunately, I can't tell you what is the difference because I do not use valgrind very often. The fact is that the problem was found when I was experimenting with CI builds (see https://github.com/kovdan01/msgpack-c/runs/2697426801). The build logs are already expired, but you can see that builds with api version 1 fail.

Reply:
Do you know how to execute only `raw` using `ctest -T memcheck` ?

Reply:
`ctest -T memcheck` doesn't report error.
But as you mentioned address sanitizer with undefined option reports error.
I accept the code change.
Thanks.

Reply:
Yes, use this:
```
ctest -T memcheck -R raw
```

Reply:
> I accept the code change.
Thanks.

Ok, thank you. I assume that I can resolve the conversation, deel free to re-open it if needed.